File: app/src/main/java/com/example/vgy_matkort/utils/HolidayImporter.kt
package com.example.vgy_matkort.utils

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Locale

data class ImportedHoliday(
    val name: String,
    val startDate: LocalDate,
    val endDate: LocalDate
)

object HolidayImporter {
    private const val VGY_URL = "https://vgy.se/lasarsdata/"
    
    /**
     * Fetches and parses holiday data from Värmdö Gymnasium's website
     * Returns only major holidays: Höstlov, Jullov, Sportlov, Påsklov
     */
    suspend fun fetchHolidaysFromVGY(): Result<List<ImportedHoliday>> = withContext(Dispatchers.IO) {
        try {
            val htmlContent = fetchWebPage(VGY_URL)
            val holidays = parseHolidays(htmlContent)
            Result.success(holidays)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    private fun fetchWebPage(urlString: String): String {
        val url = URL(urlString)
        val connection = url.openConnection() as HttpURLConnection
        connection.requestMethod = "GET"
        connection.connectTimeout = 10000
        connection.readTimeout = 10000
        
        return try {
            val reader = BufferedReader(InputStreamReader(connection.inputStream))
            reader.use { it.readText() }
        } finally {
            connection.disconnect()
        }
    }
    
    private fun parseHolidays(htmlContent: String): List<ImportedHoliday> {
        val holidays = mutableListOf<ImportedHoliday>()
        
        // Major holidays we want to import
        val majorHolidays = listOf("Höstlov", "Jullov", "Sportlov", "Påsklov")
        
        // Patterns to match:
        // "27/10 – 31/10 Höstlov"
        // "22/12 – 7/1 Jullov"
        // "23/2 – 27/2 Sportlov"
        // "3/4 – 10/4 Påsklov"
        
        val dateRangePattern = Regex("""(\d{1,2})/(\d{1,2})\s*[–-]\s*(\d{1,2})/(\d{1,2})\s+(\w+)""")
        
        dateRangePattern.findAll(htmlContent).forEach { matchResult ->
            val (startDay, startMonth, endDay, endMonth, name) = matchResult.destructured
            
            // Check if this is a major holiday
            if (majorHolidays.any { name.contains(it, ignoreCase = true) }) {
                try {
                    // Determine the year based on the month
                    val currentYear = LocalDate.now().year
                    val currentMonth = LocalDate.now().monthValue
                    
                    // If we're in autumn/winter and the holiday is in spring, it's next year
                    // If we're in spring and the holiday is in autumn/winter, it's this year or next
                    var startYear = currentYear
                    var endYear = currentYear
                    
                    val startMonthInt = startMonth.toInt()
                    val endMonthInt = endMonth.toInt()
                    
                    // Handle year transitions (e.g., Jullov from Dec to Jan)
                    if (startMonthInt > endMonthInt) {
                        endYear = startYear + 1
                    }
                    
                    // If we're past the holiday this year, use next year
                    if (currentMonth > endMonthInt || (currentMonth == endMonthInt && LocalDate.now().dayOfMonth > endDay.toInt())) {
                        startYear++
                        endYear++
                    }
                    
                    val startDate = LocalDate.of(startYear, startMonthInt, startDay.toInt())
                    val endDate = LocalDate.of(endYear, endMonthInt, endDay.toInt())
                    
                    // Clean up the name
                    val cleanName = when {
                        name.contains("Höstlov", ignoreCase = true) -> "Höstlov"
                        name.contains("Jullov", ignoreCase = true) -> "Jullov"
                        name.contains("Sportlov", ignoreCase = true) -> "Sportlov"
                        name.contains("Påsklov", ignoreCase = true) -> "Påsklov"
                        else -> name
                    }
                    
                    holidays.add(ImportedHoliday(cleanName, startDate, endDate))
                } catch (e: Exception) {
                    // Skip invalid dates
                }
            }
        }
        
        return holidays
    }
    
    /**
     * Parses holidays from CSV format
     * Expected format: name,startDate,endDate
     * Date format: YYYY-MM-DD
     */
    fun parseFromCSV(csvContent: String): List<ImportedHoliday> {
        val holidays = mutableListOf<ImportedHoliday>()
        val lines = csvContent.lines().filter { it.isNotBlank() }
        
        lines.forEach { line ->
            val parts = line.split(",").map { it.trim() }
            if (parts.size >= 3) {
                try {
                    val name = parts[0]
                    val startDate = LocalDate.parse(parts[1])
                    val endDate = LocalDate.parse(parts[2])
                    holidays.add(ImportedHoliday(name, startDate, endDate))
                } catch (e: Exception) {
                    // Skip invalid lines
                }
            }
        }
        
        return holidays
    }
}

File: iOS/MatkortCore/Sources/MatkortCore/HolidayImporter.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public struct ImportedHoliday: Equatable {
    public var name: String
    public var startDate: Date
    public var endDate: Date

    public init(name: String, startDate: Date, endDate: Date) {
        self.name = name
        self.startDate = startDate
        self.endDate = endDate
    }
}

public struct HolidayImporter {
    public var vgyURL = URL(string: "https://vgy.se/lasarsdata/")!
    public var calendar: Calendar

    public init(calendar: Calendar = .current) {
        self.calendar = calendar
    }

    public func fetchHolidaysFromVGY() async throws -> [ImportedHoliday] {
        let (data, _) = try await URLSession.shared.data(from: vgyURL)
        guard let html = String(data: data, encoding: .utf8) else { return [] }
        return parseHolidays(htmlContent: html)
    }

    public func parseHolidays(htmlContent: String, now: Date = .now) -> [ImportedHoliday] {
        let major = ["Höstlov", "Jullov", "Sportlov", "Påsklov"]
        let pattern = #"(\d{1,2})/(\d{1,2})\s*[–-]\s*(\d{1,2})/(\d{1,2})\s+([A-Za-zÅÄÖåäö]+)"#
        let regex = try? NSRegularExpression(pattern: pattern)
        let nsRange = NSRange(htmlContent.startIndex..<htmlContent.endIndex, in: htmlContent)
        let currentYear = calendar.component(.year, from: now)
        let currentMonth = calendar.component(.month, from: now)
        let currentDay = calendar.component(.day, from: now)

        var imported: [ImportedHoliday] = []
        regex?.matches(in: htmlContent, range: nsRange).forEach { match in
            guard match.numberOfRanges == 6,
                  let sDay = int(match.range(at: 1), in: htmlContent),
                  let sMonth = int(match.range(at: 2), in: htmlContent),
                  let eDay = int(match.range(at: 3), in: htmlContent),
                  let eMonth = int(match.range(at: 4), in: htmlContent),
                  let rawName = substring(match.range(at: 5), in: htmlContent) else { return }

            guard major.contains(where: { rawName.localizedCaseInsensitiveContains($0) }) else { return }

            var startYear = currentYear
            var endYear = currentYear
            if sMonth > eMonth { endYear += 1 }
            if currentMonth > eMonth || (currentMonth == eMonth && currentDay > eDay) {
                startYear += 1
                endYear += 1
            }

            guard let start = calendar.date(from: DateComponents(year: startYear, month: sMonth, day: sDay)),
                  let end = calendar.date(from: DateComponents(year: endYear, month: eMonth, day: eDay)) else { return }

            let name: String
            switch rawName.lowercased() {
            case let s where s.contains("höstlov"): name = "Höstlov"
            case let s where s.contains("jullov"): name = "Jullov"
            case let s where s.contains("sportlov"): name = "Sportlov"
            case let s where s.contains("påsklov"): name = "Påsklov"
            default: name = rawName
            }

            imported.append(ImportedHoliday(name: name, startDate: start, endDate: end))
        }

        return imported
    }

    private func int(_ range: NSRange, in text: String) -> Int? {
        substring(range, in: text).flatMap(Int.init)
    }

    private func substring(_ range: NSRange, in text: String) -> String? {
        guard let r = Range(range, in: text) else { return nil }
        return String(text[r])
    }
}
